Adding optional sound speed calculation and diagnostic

averdy opened the pull request:
## What changes does this PR introduce?
(Bug fix, feature, docs update, ...)

Adding a calculation for the speed of sound in seawater to the main code. This is meant to replace the sound speed diagnostic "csmm" currently in ecco/cost_averagesfields.F . The new global variable CSound enables the use of tomography observations in cost function calculations.

## What is the current behaviour? 
(You can also link to an open issue here)

Speed of sound is diagnosed in ecco/cost_averagesfields.F

## What is the new behaviour 
(if this is a feature change)?

Speed of sound is calculated in the new subroutine diags_sound_speed.F, which is called in dynamics.F . A new variable "csound" is available for cost function calculations (e.g. in profiles package) and cna be output as a diagnostic. 

## Does this PR introduce a breaking change? 
(What changes might users need to make in their application due to this PR?)

No, to use this one must define the flag INCLUDE_SOUNDSPEED_CALC_CODE in CPP_OPTIONS.h  (undef by default).

## Other information:

The subroutine diags_sound_speed.F was modeled after diags_phi_hyd.F, following @mjlosch 's suggestion. I am opening this PR as a way to get feedback on where would be an appropriate way to call the subroutine. It is currently called in dynamics.F but could easily be moved elsewhere. 

## Suggested addition to `tag-index`
(To avoid unnecessary merge conflicts, please don't update `tag-index`. One of the admins will do that when merging your pull request.)

- Added option for sound speed calculation in the main code

mmazloff commented:
Looks great! Thanks @averdy !

averdy commented:
@mmazloff Not sure! If you can check that would be great. Thanks!

mmazloff commented:
FYI to developers: testreport is good!

mjlosch commented on model/src/dynamics.F:
```diff
@@ -659,6 +659,11 @@ SUBROUTINE DYNAMICS(myTime, myIter, myThid)
 #endif
 
 C---+----1----+----2----+----3----+----4----+----5----+----6----+----7-|--+----|
+CAV(
```
This needs to be moved outside of the bi/bj loops (because the routine does them already. I would move it just before the diagnostics block

mjlosch commented on model/src/diags_sound_speed.F:
```diff
@@ -0,0 +1,168 @@
+C pr units: meters
+                  pr = rhoConst*(
+     &                   totPhiHyd(i,j,k,bi,bj)
+     &                  -rC(k)*gravity )*0.0001
```
with have a variable SItodBar = 1e-4 for this conversion (lives in EOS.h)

averdy replied:
@mlosch: Thank you for your comments and suggestions! 

csound was implemented as a global variable so it could be used in cost function. We plan to assimilate csound observations so other packages will need it (we're developing a new package that includes csound cost; that code will be checked in subsequently).

I moved the call to diags_sound_speed.F out of the bi/bj loops (thanks for catching that!) and used the variable SItodBar as suggested.

mjlosch reviewed (changes requested):
@averdy 
Thanks for this contribution. There are a few comments and a few things need to be changed. I guess the main question is, what do you want to do with the new 3D field `csound`. If it's just a diagnostic, is could be implemented much more economically, but you seem to have plans with this, right?

jm-c commented:
I have one comment: this sound-speed calculation is independent of which EOS is used, so it's really a diagnostic of sound-speed
as opposed to the sound-speed that one can compute from each EOS (except Poly3 & Linear). I think it's important to document it this way so that when, one day, we need to bring-in the sound-speed that is consistent with the EOS in use we don't have too much confusion.

averdy commented:
@jm-c: This is an important point; for our purpose, sound-speed derived from EOS is not accurate enough to be consistent with sound speed observations. We have clarified with comments in the header of the main subroutine and above the CPP_OPTIONS flag that this sound-speed is independent from EOS.

averdy commented:
@mjlosch 
1. Added the #include EOS.h in diags_sound_speed.F
2. I am not familiar with best practices for mitgcm code appearance and would be very grateful if you did it (and I want to learn eventually!)
3. No reason, I was copying the original code. I've changed it to be not factorized everywhere, for now (can be changed again if more efficient or more elegant to have it factorized). 
4. I've added csound profiles in our adjoint example (verification/global_oce_biogeo_bling/code_ad and input_ad). Also added the new routine to model_ad_diff.list. I did not get new recomputations. Please let me know how it's working for you!

Thanks!

mjlosch commented:
1. Forward simulations work for me, thanks 👍 
2. If you want to learn, you could use dynamics.F as an example (that's what I would do anyway): not more than one empty line at a time, upper case FORTRAN statements (INTEGER, DO, IF THEN, etc.). Give it a try, I can help afterwards.
3. I think factored is better, but maybe it would also make sense to change something like `0.000000438031096213` to `0.438031096213 _d -6`? The original matlab code has that, so why painstakingly remove it? Do we need the original matlab code in this routine?
4. AD-test works for me, so the place where you call this function looks good (after computing `totphihyd`).
The new cost function is 5 order of magnitude larger than before (and as a consequence all AD verification numbers are off entirely). Is that intentional?

averdy commented:
For the record, I did not write that piece code, just copied it from pkg/ecco/cost_averagesfields.F ! @mjlosch , let me know how it looks now. I tried to make it as clean as possible. There's no obvious way to factorize the mixed "cstp" term, but feel free to rewrite it as you see fit. 

Regarding the adjoint test, it is a completely different cost function so yes, the values are intentionally very different. I've updated results/output_adm.txt

mjlosch commented:
The file looks great, thanks! I just committed a few small tweaks, some of them are a matter of taste, I guess.
In a separate commit I attempted a factorisation. See if you like it. If not, we can just revert that.


There is is something wrong with the AD test. With `./testreport -adm -ncad -devel -t global_oce_biogeo_bling`, I get 
```
G D M    C  A  F        m  s        m  s        m  s        m  s
e p a R  o  d  D  m  m  e  .  m  m  e  .  m  m  e  .  m  m  e  .
n n k u  s  G  G  i  a  a  d  i  a  a  d  i  a  a  d  i  a  a  d
2 d e n  t  r  r  n  x  n  .  n  x  n  .  n  x  n  .  n  x  n  .

Y Y Y Y  5>22<22  3  3  3  3  3  5  5  5  3  3  3  3  3  3  3  3 22 22 22 22 22 22 22 22 pass  global_oce_biogeo_bling  (e=0, w=14)
```
and all gradient checks are zero:
```
verification (soundspeed)> grep "grdchk [o, ]" global_oce_biogeo_bling/run/output_adm.txt 
 pkg/grdchk               compiled   and   used ( useGrdchk                = T )
(PID.TID 0000.0001) grdchk output h.p:  Id Itile Jtile LAYER   bi   bj   X(Id)           X(Id)+/-EPS
(PID.TID 0000.0001) grdchk output h.c:  Id  FC                   FC1                  FC2
(PID.TID 0000.0001) grdchk output h.g:  Id     FC1-FC2/(2*EPS)      ADJ GRAD(FC)         1-FDGRD/ADGRD
(PID.TID 0000.0001) grdchk output (p):   1    34    10     1    2    1   0.000000000E+00 -1.000000000E-05
(PID.TID 0000.0001) grdchk output (c):   1  5.1587632737210E+07  5.1587632737210E+07  5.1587632737210E+07
(PID.TID 0000.0001) grdchk output (g):   1     0.0000000000000E+00  0.0000000000000E+00  0.0000000000000E+00
(PID.TID 0000.0001) grdchk output (p):   2    35    10     1    2    1   0.000000000E+00 -1.000000000E-05
(PID.TID 0000.0001) grdchk output (c):   2  5.1587632737210E+07  5.1587632737210E+07  5.1587632737210E+07
(PID.TID 0000.0001) grdchk output (g):   2     0.0000000000000E+00  0.0000000000000E+00  0.0000000000000E+00
(PID.TID 0000.0001) grdchk output (p):   3    36    10     1    2    1   0.000000000E+00 -1.000000000E-05
(PID.TID 0000.0001) grdchk output (c):   3  5.1587632737210E+07  5.1587632737210E+07  5.1587632737210E+07
(PID.TID 0000.0001) grdchk output (g):   3     0.0000000000000E+00  0.0000000000000E+00  0.0000000000000E+00
(PID.TID 0000.0001) grdchk output (p):   4    37    10     1    2    1   0.000000000E+00 -1.000000000E-05
(PID.TID 0000.0001) grdchk output (c):   4  5.1587632737210E+07  5.1587632737210E+07  5.1587632737210E+07
(PID.TID 0000.0001) grdchk output (g):   4     0.0000000000000E+00  0.0000000000000E+00  0.0000000000000E+00
(PID.TID 0000.0001) grdchk  summary  :  RMS of    4 ratios =  0.0000000000000E+00
```
That's something to sort out, I am afraid. I think you need to either figure out, why the cost function is not sensitive to the tested control parameters in gradient check and fix that, or revert to the original experiment and not test this routine now (now that we know that there are no extra recomputations incurred).

averdy commented:
The gradient check was for the sensitivity of pH to alkalinity, and not surprisingly sound speed is not sensitive to that control parameter. I'm going to revert to the original adjoint experiment and save the sound speed test for a separate verification experiment (future PR). 

Thank you for your edits, I think it looks great!

mjlosch reviewed (approved):
LGTM

averdy commented:
@jm-c Is there more that needs to be done for this to be approved? Please let me know if anything should be changed. Thanks!

jm-c reviewed (changes requested):
@averdy There are 2 issues that I can see:
1) the adjoint test for experiment global_oce_biogeo_bling is failing because the reference output has been changed (but does not need to since nothing has changed for this test).
2) one of the main purpose of having a CPP option around this diagnostic calculation is to allow  to skip  the addition of a 3-D array in common block. However, csound addition to DYNVARS.h is not protected within ifdef/endif INCLUDE_SOUNDSPEED_CALC_CODE.
To fix (2), it might be better just to add a single variable common block (rather than adding csound at the end of DYNVARS_DIAG), similar to DYNVARS_sigmaR.

Minor details:
a) just to fit in MITgcm own rules, would be good add initialisation of csound in ini_dynvars.F (like we do for any varaibles from DYNVARS.h). Even more minor, csound could be changed to "cSound" inside the code (but the diagnostics name spelling is good as it is now).
And the units of csound could be documented in DYNVARS.h (even though it's obvious). 
b) the calculation of the speed of sound might take a little bit of CPU time, so we may want to skip this when it's not needed. Currently, this would be when diagnostics "CSound  " is not used but this could change in the future. I could make some changes to diags_sound_speed.F and let you check if it's good for you (can always revert the changes).

Finally, I totally agree with your comment regarding the place in the main code where S/R DIAGS_SOUND_SPEED is called. it's in a good place to start with and could be changed if/when needed.

averdy commented:
@jm-c Thank you for your suggestions! 

1. Deleted output_adm.txt, as should have been done when we reverted to the original adjoint test experiment. 

2. In DYNVARS.h: moved cSound to its own common block, protected within ifdef/endif INCLUDE_SOUNDSPEED_CALC_CODE. 

a) Changed csound to cSound throughout the code, added initialization in ini_dynvars.F, and specified units in DYNVARS.h.

b) Sure, that sounds like a good idea. I'm interested to see how you would do it!

jm-c commented:
@averdy There was one missing ifdef ALLOW_DIAGNOSTICS in diags_sound_speed.F that I fixed (my first commit to this file).
The second commit is just to skip the sound-speed calculation when it's not used. Please take a look and let me know if it is good  for you or if there is any problem.

averdy commented:
@jm-c looks great! Thanks!

jm-c commented:
OK, I will merge this PR later today. @averdy  Thanks !
